### PR TITLE
NXP 1060 core ADU Sample

### PR DESCRIPTION
**Pull request by vaavva**
## Purpose

* Adds functionality to enable ADU on the NXP 1060 sample. Readme to come in a follow-up PR. New files within the demos/projects/NXP/mimxrt1060/nxp_code folder are from NXP. The differences in the ADU linker file can be seen easier here: https://github.com/Azure-Samples/iot-middleware-freertos-samples/compare/vaavva/nxp-adu-sample...24ff235dce23f6d5d964faea19595c9617697451#diff-d42757936b2322d52ceca2a6b22cdca7f200e1ac568df508b1f2d6d12de606f3

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

**Review by RLeclair — APPROVED** [review]

File: demos/projects/NXP/mimxrt1060/CMakeLists.txt [hunk 1]
```diff
@@ -102,6 +105,7 @@ include(driver_dcp)
 include(driver_snvs_hp)
 
 add_executable(${PROJECT_NAME} ${PROJECT_SOURCES})
+add_linker(${PROJECT_NAME})
```
Comment (ericwol-msft) [inline demos/projects/NXP/mimxrt1060/CMakeLists.txt:108]: what does this do?

File: demos/projects/NXP/mimxrt1060/config/azure_iot_config.h [hunk 2]
```diff
@@ -40,6 +40,23 @@ extern void vLoggingPrintf( const char * pcFormatString,
     #define SdkLog( message )    vLoggingPrintf message
 #endif
 
+/* Middleware logging */
```
Comment (ericwol-msft) [inline demos/projects/NXP/mimxrt1060/config/azure_iot_config.h:43]:
```suggestion
/* Middleware logging for NXP platform*/

```

File: demos/projects/NXP/mimxrt1060/gcc_flags.cmake [hunk 3]
```diff
@@ -46,9 +46,17 @@ set(CMAKE_EXE_LINKER_FLAGS ${CMAKE_EXE_LINKER_FLAGS} " \
...
 endfunction()
+
+function(add_adu_linker TARGET_NAME)
+    target_compile_options(${TARGET_NAME} PRIVATE -DSOC_REMAP_ENABLE)
+    target_link_options(${TARGET_NAME} PRIVATE -T${CMAKE_CURRENT_SOURCE_DIR}/MIMXRT1062xxxxx_flexspi_nor.ld)
+endfunction()
+
+function(add_linker TARGET_NAME)
+    target_link_options(${TARGET_NAME} PRIVATE -T${CMAKE_CURRENT_SOURCE_DIR}/MIMXRT1062xxxxx_sdram.ld)
+endfunction()
```
Comment (ericwol-msft) [inline demos/projects/NXP/mimxrt1060/gcc_flags.cmake:62]: eol

File: demos/projects/NXP/mimxrt1060/main.c [hunk 4]
```diff
@@ -292,6 +303,11 @@ void vApplicationDaemonTaskStartupHook( void )
     prvInitializeRTC();
     prvInitializeSNTP();
 
+    #ifdef ADU_SAMPLE
+        /* make the last update fully effective */
```
Comment (ericwol-msft) [inline demos/projects/NXP/mimxrt1060/main.c:307]:
```suggestion
        /* make the last flash update fully effective */
```

File: demos/projects/NXP/mimxrt1060/CMakeLists.txt (hunk 1 above)
Comment (vaavva) [reply demos/projects/NXP/mimxrt1060/CMakeLists.txt:108]: Adds the linker file for the sample - the ADU and non-ADU samples need different linker files now - any wording suggestions to make this clearer?

File: demos/projects/NXP/mimxrt1060/port/azure_iot_flash_platform.c [hunk 5]
```diff
@@ -37,81 +40,160 @@ static AzureIoTResult_t prvBase64Decode( uint8_t * base64Encoded,
         return eAzureIoTErrorFailed;
     }
 
-    LogInfo( ( "Unencoded the base64 encoding\r\n" ) );
+    AZLogInfo( ( "Unencoded the base64 encoding\r\n" ) );
```
Comment (ericwol-msft) [inline demos/projects/NXP/mimxrt1060/port/azure_iot_flash_platform.c:43]: too verbose?

File: demos/projects/NXP/mimxrt1060/port/azure_iot_flash_platform.c [hunk 6]
```diff
@@ -37,81 +40,160 @@ static AzureIoTResult_t prvBase64Decode( uint8_t * base64Encoded,
...
 {
-    ( void ) pxAduImage;
+    pxAduImage->ulCurrentOffset = 0;
+    pxAduImage->ulImageFileSize = 0;
 
-    /* TODO: Fill in initialization for NXP flash writing capability. */
+    AzureIoTResult_t xResult = eAzureIoTSuccess;
+    status_t xStatus;
+    uint8_t ucImagePosition;
+    volatile uint32_t ulPrimask;
```
Comment (ericwol-msft) [inline demos/projects/NXP/mimxrt1060/port/azure_iot_flash_platform.c:56]: volatile not really needed

File: demos/projects/NXP/mimxrt1060/port/azure_iot_flash_platform.c [hunk 7]
```diff
@@ -37,81 +40,160 @@ static AzureIoTResult_t prvBase64Decode( uint8_t * base64Encoded,
...
 
-    LogInfo( ( "AzureIoTPlatform_Init()\r\n" ) );
+    sfw_flash_read( REMAP_FLAG_ADDRESS, &ucImagePosition, 1 );
 
-    return eAzureIoTSuccess;
+    if( ucImagePosition == 0x01 )
+    {
+        pxAduImage->xUpdatePartition = FLASH_AREA_IMAGE_2_OFFSET;
+    }
+    else if( ucImagePosition == 0x02 )
```
Comment (ericwol-msft) [inline demos/projects/NXP/mimxrt1060/port/azure_iot_flash_platform.c:64]: 0x01 0x02  enums?

File: demos/projects/NXP/mimxrt1060/port/azure_iot_flash_platform.c (hunk 5 above)
Comment (vaavva) [reply demos/projects/NXP/mimxrt1060/port/azure_iot_flash_platform.c:43]: This matches the logging for the other ADU samples, so I'll leave it as is.

File: demos/projects/NXP/mimxrt1060/port/azure_iot_flash_platform.c [hunk 8]
```diff
@@ -37,81 +40,160 @@ static AzureIoTResult_t prvBase64Decode( uint8_t * base64Encoded,
...
 
-    /* TODO: Fill in to write pData of size ulBlockSize to ulOffset in memory. */
+    ulPrimask = DisableGlobalIRQ();
+    xStatus = sfw_flash_write( pucNextWriteAddr, pData, ulBlockSize );
+    EnableGlobalIRQ( ulPrimask );
 
-    LogInfo( ( "AzureIoTPlatform_WriteBlock(): offset %i\r\n", ulOffset ) );
+    if( xStatus )
+    {
+        xResult = eAzureIoTErrorFailed;
```
Comment (ericwol-msft) [inline demos/projects/NXP/mimxrt1060/port/azure_iot_flash_platform.c:108]: log error here?

File: demos/projects/NXP/mimxrt1060/port/azure_iot_flash_platform.c [hunk 9]
```diff
@@ -37,81 +40,160 @@ static AzureIoTResult_t prvBase64Decode( uint8_t * base64Encoded,
...
+    volatile uint32_t ulPrimask;
 
-    /* TODO: Fill in to verify bytes written to flash bank match the SHA256 given by pucSHA256Hash. */
+    AZLogInfo( ( "Base64 Encoded Hash from ADU: %.*s", ulSHA256HashLength, pucSHA256Hash ) );
+    xResult = prvBase64Decode( pucSHA256Hash, ulSHA256HashLength, ucDecodedManifestHash, azureiotflashSHA_256_SIZE, ( size_t * ) &ulOutputSize );
 
-    LogInfo( ( "AzureIoTPlatform_VerifyImage()\r\n" ) );
+    if( xResult != eAzureIoTSuccess )
+    {
+        AZLogError( ( "Unable to decode base64 SHA256\r\n" ) );
```
Comment (ericwol-msft) [inline demos/projects/NXP/mimxrt1060/port/azure_iot_flash_platform.c:128]:
```suggestion
        AZLogError( ( "Unable to decode image hash SHA256\r\n" ) );
```

**Review by ericwol-msft — APPROVED** [review]

**Pull request merged**